modules/packages/LinearAlgebra.chpl

```diff
@@ -560,16 +559,29 @@ proc matPow(A: [], b) where isNumeric(b) {
 
 pragma "no doc"
 /* Exponentiate by squaring recursively */
-private proc _expBySquaring(x: ?t, n): t {
-    // TODO -- _expBySquaring(pinv(x), -n);
-    if n < 0  then halt("Negative powers not yet supported");
-    else if n == 0  then return eye(x.domain);
-    else if n == 1  then return x;
-    else if n%2 == 0  then return _expBySquaring(dot(x, x), n / 2);
-    else return dot(x, _expBySquaring(dot(x, x), (n - 1) / 2));
+private proc _expBySquaring(x: ?t, n): t{
+  // TODO -- _expBySquaring(pinv(x), -n);
+  if(n<0) then halt("Negative powers not yet supported");
+  else{
+    var product = eye(x.domain);
+    var basis = x;
+    var no = n;
+    while(no>0){
+      if(no%2==1)then{
+        product = dot(product,basis);
```

Comment:
I agree, no reason to pay an allocation/initialization cost every time around.

A micro-optimization -- the first multiply for product is an identity multiply, and can probably be optimized away as well. This will matter less and less for higher powers, but will for lower powers (since the code always does that at least once).